**Ejercicio 1:** Selecciona una API pública y extrae datos utilizando Python.

Tengo como afición la astronomía y me ha parecido realizar este ejercicio con algo relacionado al tema.

Selecciono la API de la NASA, del Servicio de Detección de objetos Cercanos a la Tierra (NeoWs) que se encuentra dentro de la página https://api.nasa.gov/ para ver si puedo obtener algunos datos relacionados con objetos entre los días 17 y 18 de octubre de 2024.


In [16]:
import requests
import pandas as pd

# Hago la solicitud a  la API de la NASA 

URL = "https://api.nasa.gov/neo/rest/v1/feed?start_date=2024-10-17&end_date=2024-10-18&api_key=DEMO_KEY"
response = requests.get(URL)

if response.status_code == 200:
    data = response.json()
    print("Solicitud Exitosa")
    
else:
    print("Error en la solicitud, detalles:", response.text)




Solicitud Exitosa


Quiero saber como están estructurados los datos del objeto creado y los inspecciono mediante la función "keys"

In [22]:
data.keys()

dict_keys(['links', 'element_count', 'near_earth_objects'])

Dentro del campo near_earth_objects creo que es donde puedo encontrar los datos importantes, consulto la estructura para poder elegir los datos que me interesan para crear un dataframe con ellos.

In [25]:
# Comento la línea de código porque en github me muestra el resultado completo de una lista de 40 objetos y es una salida muy larga, a efectos en VSC me indicó la información
# que necesitaba al limitarla.

# data['near_earth_objects'] (código utilizado)

Creo la estructura para el dataframe con los datos de las columnas en español para que sean comprensibles y con los elementos que me interesan

In [19]:
# Extraigo el diccionario del objeto data que contiene la información de las fechas indicadas en la petición a la API, el diccionario de tal manera que las claves son las fechas y los valores son listas de asteriodes que se acercarán en esas fechas

near_earth_objects = data['near_earth_objects']

# inicializo una lista vacía donde se almacenarán los datos en formato diccionario, para posteriormente crear el dataframe

all_neo_data = []

# Se realiza un bucle sobre las fechas, se itera sobre cada clave o fecha y se hace un bucle sobre los asteriodes
for date in near_earth_objects:
    for neo in near_earth_objects[date]:
        neo_data = {
            'id': neo['id'],
            'nombre': neo['name'],
            'magnitud_absoluta_h': neo['absolute_magnitude_h'],
            'diametro_min_estimado_km': neo['estimated_diameter']['kilometers']['estimated_diameter_min'],
            'diametro_max_estimado_km': neo['estimated_diameter']['kilometers']['estimated_diameter_max'],
            'potencialmente_peligroso': neo['is_potentially_hazardous_asteroid'],
            'fecha_aproximacion_cercana': neo['close_approach_data'][0]['close_approach_date'],
            'velocidad_relativa_km_per_seg': neo['close_approach_data'][0]['relative_velocity']['kilometers_per_second'],
            'distancia_fallo_km': neo['close_approach_data'][0]['miss_distance']['kilometers'],
            'cuerpo_que_orbita': neo['close_approach_data'][0]['orbiting_body']
        }
        all_neo_data.append(neo_data) # Se crea el diccionario (neo_data) donde se almacena la información de cada asteroide en el diccionario vacío (all_neo_data)

# Convierto los datos en un Dataframe e imprimo el resultado

df_neo = pd.DataFrame(all_neo_data)

# Muestro el resultado

df_neo




id              nombre  magnitud_absoluta_h  \
0    2483547  483547 (2003 WM25)               17.650   
1    3137843         (2002 RD27)               18.140   
2    3145497          (2002 YO2)               22.700   
3    3529626           (2010 MA)               25.700   
4    3761647         (2016 UY35)               24.700   
5   54016334          (2020 FV3)               27.700   
6   54051583          (2020 QP6)               20.730   
7   54055117          (2020 TN3)               24.200   
8   54316690           (2022 UO)               26.530   
9   54320707         (2022 UO10)               26.760   
10  54347994          (2023 EN2)               23.120   
11  54445848          (2024 LE1)               26.310   
12  54482434         (2024 RV50)               20.970   
13  54487286          (2024 TX2)               23.820   
14  54487954          (2024 TN4)               22.570   
15  54489676          (2024 TX8)               24.230   
16  54490132          (2024 TT9)               24.200   
17  54494098         (2024 TB22)               19.323   
18   2462238   462238 (2008 CN1)               20.660   
19   3388311         (2007 TQ24)               22.060   
20   3775093         (2017 KR31)               24.100   
21   3836113          (2018 VW4)               24.520   
22   3879238           (2019 UZ)               27.800   
23   3883101          (2019 UZ7)               21.040   
24   3892267         (2019 UH14)               23.800   
25   3989721          (2020 CU1)               24.190   
26  54144232          (2021 JN2)               23.770   
27  54145782          (2021 JG9)               25.200   
28  54205435          (2021 TN6)               25.060   
29  54243523          (2022 BM5)               25.630   
30  54393421         (2023 TK12)               25.770   
31  54395029           (2023 UH)               26.130   
32  54467171         (2024 NN11)               19.310   
33  54482419          (2024 SD2)               25.650   
34  54483132          (2024 ST2)               25.140   
35  54483152          (2024 SP3)               21.840   
36  54487060          (2024 TO1)               25.620   
37  54490139         (2024 TC10)               21.550   
38  54490522         (2024 TQ11)               23.141   
39  54491553         (2024 TC15)               26.862   

    diametro_min_estimado_km  diametro_max_estimado_km  \
0                   0.784431                  1.754042   
1                   0.625972                  1.399716   
2                   0.076658                  0.171412   
3                   0.019256                  0.043057   
4                   0.030518                  0.068240   
5                   0.007666                  0.017141   
6                   0.189913                  0.424659   
7                   0.038420                  0.085909   
8                   0.013139                  0.029379   
9                   0.011818                  0.026427   
10                  0.063176                  0.141267   
11                  0.014540                  0.032512   
12                  0.170042                  0.380224   
13                  0.045767                  0.102339   
14                  0.081387                  0.181987   
15                  0.037893                  0.084731   
16                  0.038420                  0.085909   
17                  0.363040                  0.811783   
18                  0.196135                  0.438571   
19                  0.102933                  0.230165   
20                  0.040230                  0.089958   
21                  0.033155                  0.074138   
22                  0.007321                  0.016370   
23                  0.164647                  0.368163   
24                  0.046191                  0.103286   
25                  0.038597                  0.086306   
26                  0.046833                  0.104722   
27                  0.024241                  0.

Quiero saber de toda esta lista cuales son los potencialmente peligrosos

In [10]:

cantidad_peligrosos = df_neo['potencialmente_peligroso'].sum()

print(f'Cantidad de objetos potencialmente peligrosos: {cantidad_peligrosos}')

Cantidad de objetos potencialmente peligrosos: 3


Quiero saber cuáles son estos objetos

In [20]:
objetos_peligrosos = df_neo['potencialmente_peligroso'] == True

df_neo[objetos_peligrosos]

id             nombre  magnitud_absoluta_h  \
12  54482434        (2024 RV50)               20.970   
17  54494098        (2024 TB22)               19.323   
18   2462238  462238 (2008 CN1)               20.660   

    diametro_min_estimado_km  diametro_max_estimado_km  \
12                  0.170042                  0.380224   
17                  0.363040                  0.811783   
18                  0.196135                  0.438571   

    potencialmente_peligroso fecha_aproximacion_cercana  \
12                      True                 2024-10-18   
17                      True                 2024-10-18   
18                      True                 2024-10-17   

   velocidad_relativa_km_per_seg  distancia_fallo_km cuerpo_que_orbita  
12                 15.5740330005   7417329.411838163             Earth  
17                 22.0048475276  11433124.266985772             Earth  
18                 15.3649328146  68174282.335175053             Earth

Con la información de esta consulta destaco por ejemplo el objeto ***2024 TB22*** (índice 17) se considera potencialmente peligroso, explico los valores obtenidos de este objeto:

- **Magnitud absoluta:** 19.32 (la brillo del objeto, cuando menor el número más brillantes y por lo tanto más grande, ejemplo: el planeta Venus tiene una magnitud de -4,38, una estrella como Polaris, la estrella polar es de +1.97 )

- **Diámetro estimado:** entre 363 y 812 mts. De impactar en la tierra podría causar un daño regional significativo.

- **Potencialmente peligroso:** True, la NASA lo considera potencialmente peligroso.

- **Fecha de aproximación cercana:** 18 de octubre de 2024

- **Velocidad relativa:** 22 km/s velocidad extremadamente alta en relación con la tierra, en caso de impacto la energía sería considerable.

- **Distancia de fallo:** 11 millones de km. aunque esté clasificado como potencialmente peligroso, pasará a una distancia segura (más lejos que la distancia entre la tierra y la luna)

- **Trayectoria:** tiene una trayectoria que lo llevará cerca de la tierra.